src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseImplicitTypingDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,148 @@
+                || stylePreferences.HasFlag(TypingStyles.VarWherePossible) && !(isIntrinsicType || isTypeApparent);
+        }
+
+        protected override bool AnalyzeVariableDeclaration(TypeSyntax typeName, SemanticModel semanticModel, OptionSet optionSet, CancellationToken cancellationToken, out TextSpan issueSpan)
+        {
+            issueSpan = default(TextSpan);
+
+            // If it is already var, return.
+            if (typeName.IsTypeInferred(semanticModel))
+            {
+                return false;
+            }
+
+            var candidateReplacementNode = SyntaxFactory.IdentifierName("var")
+                                                .WithLeadingTrivia(typeName.GetLeadingTrivia())
+                                                .WithTrailingTrivia(typeName.GetTrailingTrivia());
+
+            var candidateIssueSpan = typeName.Span;
+
+            // If there exists a type named var, return.
+            var conflict = semanticModel.GetSpeculativeSymbolInfo(typeName.SpanStart, candidateReplacementNode, SpeculativeBindingOption.BindAsTypeOrNamespace).Symbol;
+            if (conflict != null && conflict.IsKind(SymbolKind.NamedType))
+            {
+                return false;
+            }
+
+            if (typeName.Parent.IsKind(SyntaxKind.VariableDeclaration) &&
+                typeName.Parent.Parent.IsKind(SyntaxKind.LocalDeclarationStatement, SyntaxKind.ForStatement, SyntaxKind.UsingStatement))
+            {
+                var variableDeclaration = (VariableDeclarationSyntax)typeName.Parent;
+
+                // implicitly typed variables cannot be constants.
+                var localDeclarationStatement = variableDeclaration.Parent as LocalDeclarationStatementSyntax;
+                if (localDeclarationStatement != null && localDeclarationStatement.IsConst)
+                {
+                    return false;
+                }
+
+                var variable = variableDeclaration.Variables.Single();
+                if (AnalyzeAssignment(variable.Identifier, typeName, variable.Initializer, semanticModel, optionSet, cancellationToken))
+                {
+                    issueSpan = candidateIssueSpan;
+                }
+            }
+            else if (typeName.Parent.IsKind(SyntaxKind.ForEachStatement))
+            {
+                issueSpan = candidateIssueSpan;
+            }
+
+            return issueSpan != default(TextSpan);
+        }
+
+        /// <summary>
+        /// Analyzes the assignment expression and rejects a given declaration if it is unsuitable for implicit typing.
+        /// </summary>
+        /// <returns>
+        /// false, if implicit typing cannot be used.
+        /// true, otherwise.
+        /// </returns>
+        protected override bool AnalyzeAssignment(SyntaxToken identifier, TypeSyntax typeName, EqualsValueClauseSyntax initializer, SemanticModel semanticModel, OptionSet optionSet, CancellationToken cancellationToken)
```

Comment:
Probably needs a better name that better implies the semantics of the result. E.g. `CanAssignmentBeImplicitlyTyped(...)`